src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs

```diff
@@ -88,6 +88,240 @@ public void TestAsyncOutputStream()
+            Assert.True(childLockList["parentFirstBeginOutputReadLine"].WaitSignal(WaitInMS), "Parent signal missing");
+
+            // Produce 1,2,3
+            Console.WriteLine(++counter);
+            Console.WriteLine(++counter);
+            Console.WriteLine(++counter);
+
+            // Signal 1,2,3 produced
+            Assert.True(childLockList["childProduced123"].TryAcquire(WaitInMS), "Child process should acquire 'childProduced123' lock");
+
+            // Wait call to CancelOutputRead() from parent
+            Assert.True(childLockList["parentCancelOutputRead"].WaitSignal(WaitInMS), "Signal to shutdown not come after 'WaitInMS'");
+
+            // Produce 4,5,6
+            Console.WriteLine(++counter);
+            Console.WriteLine(++counter);
+            Console.WriteLine(++counter);
+
+            // Signal 4,5,6 produced
+            Assert.True(childLockList["childProduced456"].TryAcquire(WaitInMS), "Child process should acquire 'childProduced456' lock");
+
+            // Wait second BeginOutputReadLine() from parent
+            Assert.True(childLockList["parentSecondBeginOutputReadLine"].WaitSignal(WaitInMS), "Parent signal missing");
+
+            // Produce 7,8,9
+            Console.WriteLine(++counter);
+            Console.WriteLine(++counter);
+            Console.WriteLine(++counter);
+
+            // Signal 7,8,9 produced
+            Assert.True(childLockList["childProduced789"].TryAcquire(WaitInMS), "Child process should acquire 'childProduced789' lock");
+
+            // Shutdown process
+            Assert.True(childLockList["shutdownChildProcess"].WaitSignal(WaitInMS), "Signal to shutdown not come after 'WaitInMS'");
+
+            return SuccessExitCode;
+        }
+
+        private Dictionary<string, FileSystemLock> TestAsyncOutputStream_BeginCancelBegin_OutputReadLine_LockHelper(string directoryName)
+        {
+            return new Dictionary<string, FileSystemLock>
+            {
+                { "childStarted", new FileSystemLock("childStarted", directoryName) },
+                { "parentFirstBeginOutputReadLine", new FileSystemLock("parentFirstBeginOutputReadLine", directoryName) },
+                { "parentSecondBeginOutputReadLine", new FileSystemLock("parentSecondBeginOutputReadLine", directoryName) },
+                { "parentCancelOutputRead", new FileSystemLock("parentCancelOutputRead", directoryName)},
+                { "childProduced123", new FileSystemLock("childProduced123", directoryName) },
+                { "childProduced456", new FileSystemLock("childProduced456", directoryName) },
+                { "childProduced789", new FileSystemLock("childProduced789", directoryName) },
+                { "shutdownChildProcess", new FileSystemLock("shutdownChildProcess", directoryName) }
+            };
+        }
+
+        /// <summary>
+        /// We save lock files in temp directory
+        /// </summary>
+        private class FileSystemLock : IDisposable
```

Comment:
Could anonymous or named pipes be used instead of a custom primitive like this? A signaler can write a but to the pipe and a waiter can read a byte.

Reply:
ok I'll try with pipes!Thank's!

Reply:
@stephentoub PTAL I used pipes.